File: .deprecated/inventory-service/src/routes/vendors.ts
import { Router, Request, Response } from 'express';
import { z } from 'zod';
import { prisma } from '@copio/core';

const router = Router();

// =============================================================================
// SCHEMAS
// =============================================================================

const CreateVendorSchema = z.object({
  // Identity
  name: z.string().min(1),
  legalName: z.string().optional(),
  code: z.string().min(1).optional(),
  taxId: z.string().optional(),
  website: z.string().url().optional().or(z.literal('')),
  
  // Classification
  tier: z.enum(['STRATEGIC', 'PREFERRED', 'STANDARD', 'PROBATION', 'SUSPENDED']).optional(),
  category: z.string().optional(),
  tags: z.array(z.string()).optional(),
  
  // Address
  address: z.record(z.unknown()).optional(),
  billingAddress: z.record(z.unknown()).optional(),
  
  // Ordering & Terms
  leadTimeDays: z.number().int().positive().optional(),
  minOrderValue: z.number().positive().optional(),
  paymentTerms: z.string().optional(),
  creditLimit: z.number().positive().optional(),
  currency: z.string().length(3).optional(),
  
  // Account Management
  accountManagerName: z.string().optional(),
  accountManagerEmail: z.string().email().optional().or(z.literal('')),
  preferredContactMethod: z.string().optional(),
  
  // Contract & Compliance
  contractStartDate: z.string().datetime().optional(),
  contractEndDate: z.string().datetime().optional(),
  insuranceExpiry: z.string().datetime().optional(),
  w9OnFile: z.boolean().optional(),
  
  // Notes
  notes: z.string().optional(),
  internalNotes: z.string().optional(),
});

const UpdateVendorSchema = CreateVendorSchema.partial().extend({
  isActive: z.boolean().optional(),
  onboardedAt: z.string().datetime().optional(),
}).refine(data => Object.keys(data).length > 0, {
  message: 'At least one field must be provided',
});

// =============================================================================
// VENDOR ROUTES
// =============================================================================

// Create vendor
router.post('/', async (req: Request, res: Response) => {
  const parsed = CreateVendorSchema.safeParse(req.body);
  if (!parsed.success) {
    return res.status(400).json({ error: parsed.error.errors });
  }

  try {
    const data = {
      ...parsed.data,
      website: parsed.data.website || undefined,
      accountManagerEmail: parsed.data.accountManagerEmail || undefined,
      contractStartDate: parsed.data.contractStartDate ? new Date(parsed.data.contractStartDate) : undefined,
      contractEndDate: parsed.data.contractEndDate ? new Date(parsed.data.contractEndDate) : undefined,
      insuranceExpiry: parsed.data.insuranceExpiry ? new Date(parsed.data.insuranceExpiry) : undefined,
    };

    const vendor = await prisma.vendor.create({
      data,
      include: {
        contacts: { where: { isActive: true }, orderBy: { isPrimary: 'desc' } },
        _count: { select: { products: true, purchaseOrders: true } },
      },
    });
    res.status(201).json(vendor);
  } catch (error: unknown) {
    if (error && typeof error === 'object' && 'code' in error && error.code === 'P2002') {
      return res.status(409).json({ error: 'Vendor code already exists' });
    }
    throw error;
  }
});

// List vendors
router.get('/', async (req: Request, res: Response) => {
  const { active, search, tier } = req.query;
  
  const where: Record<string, unknown> = {};
  if (active === 'true') where.isActive = true;
  if (active === 'false') where.isActive = false;
  if (tier) where.tier = tier;
  if (search) {
    where.OR = [
      { name: { contains: search as string, mode: 'insensitive' } },
      { code: { contains: search as string, mode: 'insensitive' } },
      { legalName: { contains: search as string, mode: 'insensitive' } },
    ];
  }

  const vendors = await prisma.vendor.findMany({
    where,
    include: {
      contacts: { where: { isActive: true, isPrimary: true }, take: 1 },
      _count: { select: { products: true, purchaseOrders: true } },
    },
    orderBy: { name: 'asc' },
  });
  res.json(vendors);
});

// Get single vendor
router.get('/:id', async (req: Request, res: Response) => {
  const vendor = await prisma.vendor.findUnique({
    where: { id: req.params.id },
    include: {
      contacts: { where: { isActive: true }, orderBy: [{ isPrimary: 'desc' }, { role: 'asc' }] },
      addresses: { where: { isActive: true }, orderBy: [{ isPrimary: 'desc' }, { type: 'asc' }] },
      documents: { where: { isActive: true }, orderBy: [{ type: 'asc' }, { uploadedAt: 'desc' }] },
      products: {
        include: {
          product: { select: { id: true, sku: true, name: true } },
        },
        orderBy: { createdAt: 'desc' },
      },
      _count: { select: { purchaseOrders: true } },
    },
  });
  if (!vendor) {
    return res.status(404).json({ error: 'Vendor not found' });
  }
  res.json(vendor);
});

// Update vendor
router.patch('/:id', async (req: Request, res: Response) => {
  const parsed = UpdateVendorSchema.safeParse(req.body);
  if (!parsed.success) {
    return res.status(400).json({ error: parsed.error.errors });
  }

  const existing = await prisma.vendor.findUnique({ where: { id: req.params.id } });
  if (!existing) {
    return res.status(404).json({ error: 'Vendor not found' });
  }

  try {
    const data = {
      ...parsed.data,
      website: parsed.data.website === '' ? null : parsed.data.website,
      accountManagerEmail: parsed.data.accountManagerEmail === '' ? null : parsed.data.accountManagerEmail,
      contractStartDate: parsed.data.contractStartDate ? new Date(parsed.data.contractStartDate) : undefined,
      contractEndDate: parsed.data.contractEndDate ? new Date(parsed.data.contractEndDate) : undefined,
      insuranceExpiry: parsed.data.insuranceExpiry ? new Date(parsed.data.insuranceExpiry) : undefined,
      onboardedAt: parsed.data.onboardedAt ? new Date(parsed.data.onboardedAt) : undefined,
    };

    const vendor = await prisma.vendor.update({
      where: { id: req.params.id },
      data,
      include: {
        contacts: { where: { isActive: true }, orderBy: { isPrimary: 'desc' } },
        _count: { select: { products: true, purchaseOrders: true } },
      },
    });
    res.json(vendor);
  } catch (error: unknown) {
    if (error && typeof error === 'object' && 'code' in error && error.code === 'P2002') {
      return res.status(409).json({ error: 'Vendor code already exists' });
    }
    throw error;
  }
});

// Delete vendor
router.delete('/:id', async (req: Request, res: Response) => {
  const existing = await prisma.vendor.findUnique({
    where: { id: req.params.id },
    include: { _count: { select: { products: true, purchaseOrders: true } } },
  });
  if (!existing) {
    return res.status(404).json({ error: 'Vendor not found' });
  }
  if (existing._count.purchaseOrders > 0) {
    return res.status(400).json({ error: 'Cannot delete vendor with purchase orders' });
  }

  // Delete contacts, vendor products, then vendor
  await prisma.vendorContact.deleteMany({ where: { vendorId: req.params.id } });
  await prisma.vendorProduct.deleteMany({ where: { vendorId: req.params.id } });
  await prisma.vendor.delete({ where: { id: req.params.id } });
  res.status(204).send();
});

// =============================================================================
// VENDOR CONTACTS (nested under vendor)
// =============================================================================

// List vendor contacts
router.get('/:id/contacts', async (req: Request, res: Response) => {
  const vendor = await prisma.vendor.findUnique({ where: { id: req.params.id } });
  if (!vendor) {
    return res.status(404).json({ error: 'Vendor not found' });
  }

  const contacts = await prisma.vendorContact.findMany({
    where: { vendorId: req.params.id },
    orderBy: [{ isPrimary: 'desc' }, { role: 'asc' }, { name: 'asc' }],
  });
  res.json(contacts);
});

// Add contact to vendor
router.post('/:id/contacts', async (req: Request, res: Response) => {
  const { name, title, email, phone, mobile, role, isPrimary, notes } = req.body;

  if (!name) {
    return res.status(400).json({ error: 'name is required' });
  }

  const vendor = await prisma.vendor.findUnique({ where: { id: req.params.id } });
  if (!vendor) {
    return res.status(404).json({ error: 'Vendor not found' });
  }

  // If setting as primary, unset other primary contacts
  if (isPrimary) {
    await prisma.vendorContact.updateMany({
      where: { vendorId: req.params.id, isPrimary: true },
      data: { isPrimary: false },
    });
  }

  const contact = await prisma.vendorContact.create({
    data: {
      vendorId: req.params.id,
      name,
      title,
      email,
      phone,
      mobile,
      role: role || 'GENERAL',
      isPrimary: isPrimary || false,
      notes,
    },
  });
  res.status(201).json(contact);
});

// =============================================================================
// QUICK STATS
// =============================================================================

router.get('/stats/summary', async (_req: Request, res: Response) => {
  const [total, byTier, expiringSoon] = await Promise.all([
    prisma.vendor.count({ where: { isActive: true } }),
    prisma.vendor.groupBy({
      by: ['tier'],
      where: { isActive: true },
      _count: true,
    }),
    prisma.vendor.count({
      where: {
        isActive: true,
        contractEndDate: {
          lte: new Date(Date.now() + 30 * 24 * 60 * 60 * 1000), // 30 days
          gte: new Date(),
        },
      },
    }),
  ]);

  res.json({
    total,
    byTier: byTier.reduce((acc, t) => ({ ...acc, [t.tier]: t._count }), {}),
    contractsExpiringSoon: expiringSoon,
  });
});

export { router as vendorsRouter };

File: packages/api/src/api/v1/vendors.py
"""Vendors API endpoints."""

from datetime import datetime, timedelta
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, Query, status
from sqlalchemy import select, func, or_
from sqlalchemy.orm import selectinload

from src.api.deps import DbSession, CurrentUser
from src.auth.dependencies import require_permission
from src.db.models import (
    Vendor,
    VendorContact,
    VendorProduct,
    PurchaseOrder,
    VendorTier,
)
from src.schemas.vendor import (
    VendorCreate,
    VendorUpdate,
    VendorRead,
    VendorListItem,
    VendorStatsSummary,
    VendorContactCreate,
    VendorContactRead,
)

router = APIRouter()


# =============================================================================
# VENDOR CRUD
# =============================================================================

@router.post("/", response_model=VendorRead, status_code=status.HTTP_201_CREATED)
async def create_vendor(
    data: VendorCreate,
    session: DbSession,
    current_user: CurrentUser,
    _perm=Depends(require_permission("vendors:create")),
):
    """Create a new vendor."""
    # Check for duplicate code
    if data.code:
        existing = await session.execute(
            select(Vendor).where(Vendor.code == data.code)
        )
        if existing.scalar_one_or_none():
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="Vendor code already exists",
            )

    vendor = Vendor(
        tenant_id=current_user.tenant_id,
        **data.model_dump(by_alias=False, exclude_none=True),
    )
    session.add(vendor)
    await session.commit()

    # Reload with relationships
    result = await session.execute(
        select(Vendor)
        .options(
            selectinload(Vendor.contacts),
            selectinload(Vendor.addresses),
            selectinload(Vendor.documents),
            selectinload(Vendor.products),
        )
        .where(Vendor.id == vendor.id)
    )
    return result.scalar_one()


@router.get("/", response_model=list[VendorListItem])
async def list_vendors(
    session: DbSession,
    current_user: CurrentUser,
    active: bool | None = Query(None),
    tier: VendorTier | None = None,
    search: str | None = None,
    _perm=Depends(require_permission("vendors:view")),
):
    """List vendors with optional filtering."""
    query = select(Vendor).where(Vendor.tenant_id == current_user.tenant_id)

    if active is not None:
        query = query.where(Vendor.is_active == active)
    if tier:
        query = query.where(Vendor.tier == tier)
    if search:
        search_filter = or_(
            Vendor.name.ilike(f"%{search}%"),
            Vendor.code.ilike(f"%{search}%"),
            Vendor.legal_name.ilike(f"%{search}%"),
        )
        query = query.where(search_filter)

    query = query.options(
        selectinload(Vendor.contacts),
    ).order_by(Vendor.name)

    result = await session.execute(query)
    vendors = result.scalars().all()

    # Get counts for each vendor
    vendor_list = []
    for vendor in vendors:
        # Get product and PO counts
        product_count = await session.execute(
            select(func.count()).where(VendorProduct.vendor_id == vendor.id)
        )
        po_count = await session.execute(
            select(func.count()).where(PurchaseOrder.vendor_id == vendor.id)
        )
        
        vendor_item = VendorListItem(
            id=vendor.id,
            name=vendor.name,
            code=vendor.code,
            tier=vendor.tier,
            is_active=vendor.is_active,
            lead_time_days=vendor.lead_time_days,
            contacts=[VendorContactRead.model_validate(c) for c in vendor.contacts if c.is_active and c.is_primary][:1],
            product_count=product_count.scalar_one(),
            po_count=po_count.scalar_one(),
        )
        vendor_list.append(vendor_item)

    return vendor_list


@router.get("/stats/summary", response_model=VendorStatsSummary)
async def get_vendor_stats(
    session: DbSession,
    current_user: CurrentUser,
    _perm=Depends(require_permission("vendors:view")),
):
    """Get vendor statistics summary."""
    # Total active vendors
    total_result = await session.execute(
        select(func.count()).where(
            Vendor.tenant_id == current_user.tenant_id,
            Vendor.is_active == True,
        )
    )
    total = total_result.scalar_one()

    # By tier
    tier_result = await session.execute(
        select(Vendor.tier, func.count())
        .where(
            Vendor.tenant_id == current_user.tenant_id,
            Vendor.is_active == True,
        )
        .group_by(Vendor.tier)
    )
    by_tier = {str(tier.value): count for tier, count in tier_result.all()}

    # Contracts expiring in 30 days
    expiring_soon = await session.execute(
        select(func.count()).where(
            Vendor.tenant_id == current_user.tenant_id,
            Vendor.is_active == True,
            Vendor.contract_end_date >= datetime.now(),
            Vendor.contract_end_date <= datetime.now() + timedelta(days=30),
        )
    )

    return VendorStatsSummary(
        total=total,
        by_tier=by_tier,
        contracts_expiring_soon=expiring_soon.scalar_one(),
    )


@router.get("/{vendor_id}", response_model=VendorRead)
async def get_vendor(
    vendor_id: UUID,
    session: DbSession,
    current_user: CurrentUser,
    _perm=Depends(require_permission("vendors:view")),
):
    """Get a single vendor with all relationships."""
    result = await session.execute(
        select(Vendor)
        .options(
            selectinload(Vendor.contacts),
            selectinload(Vendor.addresses),
            selectinload(Vendor.documents),
            selectinload(Vendor.products).selectinload(VendorProduct.product),
        )
        .where(
            Vendor.id == vendor_id,
            Vendor.tenant_id == current_user.tenant_id,
        )
    )
    vendor = result.scalar_one_or_none()
    if not vendor:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Vendor not found",
        )
    return vendor


@router.patch("/{vendor_id}", response_model=VendorRead)
async def update_vendor(
    vendor_id: UUID,
    data: VendorUpdate,
    session: DbSession,
    current_user: CurrentUser,
    _perm=Depends(require_permission("vendors:edit")),
):
    """Update a vendor."""
    result = await session.execute(
        select(Vendor).where(
            Vendor.id == vendor_id,
            Vendor.tenant_id == current_user.tenant_id,
        )
    )
    vendor = result.scalar_one_or_none()
    if not vendor:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Vendor not found",
        )

    # Check for duplicate code
    if data.code and data.code != vendor.code:
        existing = await session.execute(
            select(Vendor).where(
                Vendor.code == data.code,
                Vendor.id != vendor_id,
            )
        )
        if existing.scalar_one_or_none():
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail="Vendor code already exists",
            )

    update_data = data.model_dump(by_alias=False, exclude_none=True)
    # Handle empty strings for optional fields
    for field in ["website", "account_manager_email"]:
        if field in update_data and update_data[field] == "":
            update_data[field] = None

    for field, value in update_data.items():
        setattr(vendor, field, value)

    await session.commit()

    # Reload with relationships
    result = await session.execute(
        select(Vendor)
        .options(
            selectinload(Vendor.contacts),
            selectinload(Vendor.addresses),
            selectinload(Vendor.documents),
            selectinload(Vendor.products),
        )
        .where(Vendor.id == vendor_id)
    )
    return result.scalar_one()


@router.delete("/{vendor_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_vendor(
    vendor_id: UUID,
    session: DbSession,
    current_user: CurrentUser,
    _perm=Depends(require_permission("vendors:delete")),
):
    """Delete a vendor."""
    result = await session.execute(
        select(Vendor).where(
            Vendor.id == vendor_id,
            Vendor.tenant_id == current_user.tenant_id,
        )
    )
    vendor = result.scalar_one_or_none()
    if not vendor:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Vendor not found",
        )

    # Check for purchase orders
    po_count = await session.execute(
        select(func.count()).where(PurchaseOrder.vendor_id == vendor_id)
    )
    if po_count.scalar_one() > 0:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Cannot delete vendor with purchase orders",
        )

    await session.delete(vendor)
    await session.commit()


# =============================================================================
# VENDOR CONTACTS (Nested under vendors)
# =============================================================================

@router.get("/{vendor_id}/contacts", response_model=list[VendorContactRead])
async def list_vendor_contacts(
    vendor_id: UUID,
    session: DbSession,
    current_user: CurrentUser,
):
    """List contacts for a vendor."""
    # Verify vendor exists
    result = await session.execute(
        select(Vendor).where(
            Vendor.id == vendor_id,
            Vendor.tenant_id == current_user.tenant_id,
        )
    )
    if not result.scalar_one_or_none():
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Vendor not found",
        )

    result = await session.execute(
        select(VendorContact)
        .where(VendorContact.vendor_id == vendor_id)
        .order_by(VendorContact.is_primary.desc(), VendorContact.role, VendorContact.name)
    )
    return result.scalars().all()


@router.post("/{vendor_id}/contacts", response_model=VendorContactRead, status_code=status.HTTP_201_CREATED)
async def create_vendor_contact(
    vendor_id: UUID,
    data: VendorContactCreate,
    session: DbSession,
    current_user: CurrentUser,
):
    """Add a contact to a vendor."""
    # Verify vendor exists
    result = await session.execute(
        select(Vendor).where(
            Vendor.id == vendor_id,
            Vendor.tenant_id == current_user.tenant_id,
        )
    )
    if not result.scalar_one_or_none():
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Vendor not found",
        )

    # If setting as primary, unset other primary contacts
    if data.is_primary:
        await session.execute(
            select(VendorContact)
            .where(
                VendorContact.vendor_id == vendor_id,
                VendorContact.is_primary == True,
            )
        )
        result = await session.execute(
            select(VendorContact).where(
                VendorContact.vendor_id == vendor_id,
                VendorContact.is_primary == True,
            )
        )
        for contact in result.scalars():
            contact.is_primary = False

    contact = VendorContact(
        vendor_id=vendor_id,
        **data.model_dump(by_alias=False),
    )
    session.add(contact)
    await session.commit()
    await session.refresh(contact)
    return contact
